### PR TITLE
Merge 1.12 to 2.0

### PR DESCRIPTION
**Pull request by juho-jaakkola**
I'm not completely sure if I fixed all the conflicts correctly, so might be best if someone took a closer look before accepting.

Conflicts:
- engine/lib/elgglib.php
- engine/lib/views.php
- engine/start.php
- views/default/css/elgg.php

### CODE REVIEW

File: engine/classes/Elgg/Application.php [hunk 1]
```diff
@@ -333,6 +333,15 @@ public function __get($name) {
...
 	private static function create() {
+		// we need to register for shutdown before Symfony registers the
+		// session_write_close() function. https://github.com/Elgg/Elgg/issues/9243
+		register_shutdown_function(function () {
+			// There are cases where we may exit before this function is defined.
+			if (function_exists('_elgg_shutdown_hook')) {
+				_elgg_shutdown_hook();
+			}
+		});
+
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application.php]: @mrclay Is this how you intended 2.0 to handle https://github.com/Elgg/Elgg/blob/1.12/engine/start.php#L31

**Comment by juho-jaakkola** [comment]
https://github.com/Elgg/Elgg/commit/c0892d774534f68f4771a6eccb1e570e5f108df0#diff-dc34c2b06846278bbedf8ab656094d59R1577

I didn't find where the `elgg_views_handle_deprecated_views()` was removed for 2.0, but this change isn't apparently needed in any case?

File: engine/classes/Elgg/Application.php [hunk 2]
```diff
@@ -333,6 +333,15 @@ public function __get($name) {
 	 * @return self
 	 */
 	private static function create() {
+		// we need to register for shutdown before Symfony registers the
+		// session_write_close() function. https://github.com/Elgg/Elgg/issues/9243
+		register_shutdown_function(function () {
+			// There are cases where we may exit before this function is defined.
+			if (function_exists('_elgg_shutdown_hook')) {
```
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application.php]: hm... is this check still needed

File: engine/classes/Elgg/Application.php (hunk 1 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: Move inside the if block, where the new instance is created.

File: engine/classes/Elgg/Application.php (hunk 2 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: Yes. You can create the app without loading core functions, I think.

File: engine/classes/Elgg/Application.php (hunk 1 above)
Comment (juho-jaakkola) [inline engine/classes/Elgg/Application.php]: Fixed

**Pull request merged**